linkerd/proxy/transport/src/listen.rs

```diff
@@ -79,68 +63,42 @@ impl<A: OrigDstAddr> Bind<A> {
     pub fn keepalive(&self) -> Option<Duration> {
         self.keepalive
     }
-}
-
-impl<O: OrigDstAddr> listen::Bind for Bind<O> {
-    type Connection = Connection;
-    type Listen = Listen<O>;
-
-    fn bind(&self) -> std::io::Result<Listen<O>> {
-        let tcp = std::net::TcpListener::bind(self.bind_addr)?;
-        let listen_addr = tcp.local_addr()?;
-        Ok(Listen {
-            listen_addr,
-            keepalive: self.keepalive,
-            orig_dst_addr: self.orig_dst_addr.clone(),
-            state: State::Init(Some(tcp)),
-        })
-    }
-}
 
-impl<O> listen::Listen for Listen<O>
-where
-    O: OrigDstAddr,
-{
-    type Connection = Connection;
-    type Error = std::io::Error;
-
-    fn listen_addr(&self) -> SocketAddr {
-        self.listen_addr
-    }
+    pub fn bind<S: Future>(
+        &self,
+        shutdown: S,
+    ) -> std::io::Result<(SocketAddr, impl Stream<Item = io::Result<Connection>>)> {
+        let listen = std::net::TcpListener::bind(self.bind_addr)?;
+        let addr = listen.local_addr()?;
+        let keepalive = self.keepalive;
+        let get_orig = self.orig_dst_addr.clone();
+
+        let accept = try_stream! {
+            // The tokio listener is built lazily so that it is initialized on
+            // the proper runtime.
+            let listen = tokio::net::TcpListener::from_std(listen).expect("listener must be valid");
+
+            futures::pin_mut!(listen);
+            futures::pin_mut!(shutdown);
```

Comment:
nit/TIOLI: i believe the `tokio::pin!` macro's `let` form (second example [here](https://docs.rs/tokio/0.2.21/tokio/macro.pin.html#examples) might let us make this look a little nicer:
```suggestion
            tokio::pin! {
                // The tokio listener is built lazily so that it is initialized on
                // the proper runtime.
                let listen = tokio::net::TcpListener::from_std(listen).expect("listener must be valid");
                let shutdown = shutdown;
            }
```